# Exploring Ebay Car Sales
In this project we'll be looking at a data set of used cars listings from a German Ebay classifieds section. The data set was originally scraped and uploaded to [kaggle](https://www.kaggle.com/orgesleka/used-cars-database), but the data set provided by dataquest for this project is a slightly modified version:
- A sample of 50,000 was taken from the data set to ensure the code running smoothly. This should be large enough of a sample to draw meaningful conclusions.
- The data set has been made more dirty for pedagogical reasons, so that it more closely resembles what you can expect from a scraped data set.

The aim of this project is to clean the data to make it available for analysis. Following this some simple analysis of the data set will be made.

We'll start by importing the relevant libraries as well as reading the data set into pandas:

In [1]:
import numpy as np
import pandas as pd
autos = pd.read_csv('autos.csv', encoding='Latin-1')

To start off we'll want to see what we're dealing with. Let's print some info about the dataframe as well as print the first few rows:

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

To confirm our previous description of the data set, it consists of 50,000 entries. There are 20 columns and they either contain integer or object (string) data.

Columns that have a "Non-Null" value lower than 50,000 includes null data. We have five such columns (vehicleType, gearbox, model, fuelType and notRepairedDamage). However, in all of these columns the share of null-values never exceed 20%.

## Column name cleaning
Note that column names use camel case (e.g fuelType) instead of the Python preferred snake case (e.g fuel_type). Let's convert the column names to snake case and make some of them more descriptive while we're at it.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Let's make our changes:

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Initial data exploration
It's time to start exploring the data. We will look for:
- text columns where almost all values are identical, as they often don't make useful objects for analysis and can be dropped.
- numerical data stored as strings. Cleaning such data and converting it to a numerical data type make it available for quantitative analysis.
- Strange behavior or anomalies that demands further investigation.

Let's use the [describe](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html)-function for both numerical and categorical columns, to quickly gain insights.

In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Initial observations:
1. The `num_photos` column looks strange.
2. In the `seller` as well as `offer_type` columns, all except one entry have the same value.
3. The `price` as well as `odometer` columns are of a numerical character, but are stored as strings as they contain their corresponding units. If the units are consistent over the entire data set, we can remove them and turn the columns into numerical data types.

Let's look closer at the strange `num_photos` column:

In [6]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

It appears as though every entry has a 0 value for this column. This doesn't contribute to our analysis, so let's drop this column along with the forementioned `seller` and `offer_type` columns that contains almost identical values for all entries:

In [7]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

Let's now look handle observation 3 and clean and convert this numerical data contained in the `price` and `odometer` columns to a numerical data type:

In [8]:
autos["price"] = autos["price"].str.replace("$","")\
                    .str.replace(",","").astype(int)
autos.rename({'price': 'price_dollar'}, axis=1, inplace=True)
autos["price_dollar"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_dollar, dtype: int32

In [9]:
autos["odometer"] = autos["odometer"].str.replace("km","")\
                    .str.replace(",","").astype(int)
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

## Exploring Odometer

In [10]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The values appear rounded and there are only 13 unique ones. This indicates that the sellers have to choose from a set of values decided on by the website. Every value contains fewer entries than those larger, with the extreme of 150,000km containing 64.85% of the entries. This indicates that more high mileage vehicles are being listed as opposed to low ones.

## Exploring Price

In [11]:
print("Unique prices:", autos['price_dollar'].unique().shape[0])
print(autos['price_dollar'].describe())
autos["price_dollar"].value_counts().head(20)

Unique prices: 2357
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price_dollar, dtype: int64

At a first glance, the prices appear rounded as they are divisable by 50. However, the '999' entry disproves this and the fact that so many of the high frequency values are divisable by 50 could be a result of people often tending to price this way. Also, the fact that there are as many as 2357 unique values among the 50,000 entries also indicate that rounding does not occur.

The most frequent entry undoubtedly is '0'. Since Ebay is an auction site this would essentially mean that these auctions potentially can be won for free. This is a strange concept and we may consider removing such entries from the data.

The maximum price is 100,000,000$ which seems quite unreasonable, so let's take a look at some of the largest:

In [12]:
autos["price_dollar"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_dollar, dtype: int64

There are some really high values in this data. However, certain cars can be very valuable, even when used.

Let's also look at the lowest prices:

In [13]:
autos["price_dollar"].value_counts().sort_index().head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_dollar, dtype: int64

As stated earlier, 0-bids is a weird concept and we probably don't want to include such entries in the data. The unrealisticly high prices, let's say those above 350,000$, will act as outliers that could skew the data. Let's remove these entries from our data:

In [14]:
autos = autos[autos["price_dollar"].between(1,350000)]
autos.describe()

price_dollar  registration_year      power_ps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        48565.000000  48565.000000  
mean             5.782251  50975.745207  
std              3.685595  25746.968398  
min              0.000000   1067.000000  
25%              3.000000  30657.000000  
50%              6.000000  49716.000000  
75%              9.000000  71665.000000  
max             12.000000  99998.000000

What we have left is what seems like a more reasonable data set, that is probably more well suited for analysis.

## Exploring date columns

There are five columns that represent dates. Some are added by the crawler and some from the website itself. Some columns have data stored as integers whereas some strings, as is apparent from the output of the info-function earlier in this project. These are the following five columns:
- `date_crawled`: added by the crawler. Stored as string.
- `last_seen`: added by the crawler. Stored as string.
- `ad_created`: from the website. Stored as string.
- `registration_month`: from the website. Stored as int.
- `registration_year`: from the website. Stored as int.

Let's convert the three columns with data represented as strings to a numerical representation to enable quantitative analysis. We'll start by looking at a few rows of these columns to understand the format presented:

In [15]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The first 10 characters represent the date. It would be interesting to gain some understanding of the distribution of dates so let's extract the dates, count the values and express them as percentages and finish by sorting them.

In [16]:
(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False)
.sort_index())

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The values range from 2016-03-05 to 2016-04-07 and every date in this interval is represented. There seems to roughly be an even distribution over all of these dates.

In [17]:
(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False)
.sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There are 76 dates represented in this column, and the range is fairly large, with the oldest dating back about 9 months from the listing date. Though we don't see many values, it seems as though the frequency increases among the more resent dates. Let's look at more rows to confirm this:

In [18]:
ad_created_view = (autos['ad_created'].str[:10]
                   .value_counts(normalize=True, dropna=False).sort_index())
print(ad_created_view.head(20))
print('\n')
print(ad_created_view.tail(20))

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
Name: ad_created, dtype: float64


2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21    0.037579
2016-03-22    0.032801
2016-03-23    0.032060
2016-03-24    0.029280
2016-03-25    0.031751
2016-03-26    0.032266
2016-03-27    0.030989
2016-03-28    0.034984
2016-03-29    0.034037
2016-03-30    0.033501
2016-03-31    0.031875
2016-04-01    0.033687
2016-04-02    0.035149
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, dtype: float64


As suspected, there is a major concentration of values among resent dates. The latest 1-2 months appears to be where a predominant share of ads were created.

In [19]:
(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False)
.sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

These values represent the date the crawler saw any listing, which gives us a rough idea of on what day the listing was removed (and likely the car sold).

There seems to roughly be an even distribution, with the exception of the last three dates where there seems to be a large concentration of values. It seems improbable that there was a huge spike in sales on these specific dates. Rather, it most likely has to do with the crawling ending after these dates, recording a high concentration of last seen around that time.

Let's also briefly look at the registration year:

In [20]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Here we can notice a couple of strange things. The earliest registration date is year 1000 and the latest is year 9999. These are obvisously not correct and this indicates that there are problems with this data.

## Incorrect Registration Year data
As we just mentioned, the registration year data includes values that are way beyond the possible in this context. These outliers lowers the correctness of the data and conclusions being drawn from it. We need to attempt to deal with this.

Since the crawling is taking place in 2016, not registration year should exceed that. The earlist year to include is a bit trickier to determine. It seems probable that it should be around the first couple of decades of the 1900s. To be on the safe side, let's include everything from year 1900.

Let's first take a look at the share of entries that would be excluded with the forementioned interval cut-off and based on that decide how to handle the situation:

In [21]:
p = 1 - (autos['registration_year'].between(1900,2016).sum())/autos.shape[0]
print("{:.5f}".format(p))

0.03879


Slightly less than 4% are outside of this "legal" interval. Since the share is so low, let's just remove the corresponding rows:

In [22]:
autos = autos[autos['registration_year'].between(1900,2016)]

With those removed, let's take a look at the distribution by looking at the 15 most populated years:

In [23]:
autos['registration_year'].value_counts(normalize=True).head(15)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
Name: registration_year, dtype: float64

It appears most vehicles are registered within a rough 15 year period between late 90s and early 2010s.

## Exploring Price by Brand

In [24]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

The top 5 makes up about 60% of the ads and is entirely made up of German brands, which is not surprising considering that the ads are collected from a German website as well as the fact that Germany has a massive auto industry.

There are a lot of brands present with very low representation. To limit our scope somewhat we will focus on the ones with over 2.5% share.

In [25]:
brand_counts = autos['brand'].value_counts(normalize=True)
popular_brands = brand_counts[brand_counts > 0.025].index
print(popular_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat'],
      dtype='object')


In [26]:
mean_prices = {}
for curr_brand in popular_brands:
    brand_prices = autos.loc[autos['brand'] == curr_brand, 'price_dollar']
    mean_prices[curr_brand] = int(brand_prices.mean())

mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813}

There are significant differences with regards to mean price between the top brands. 
- Three brands really stand out as expensive; audi, mercedes_benz and bwm, with a mean price of 9336, 8628 and 8332 respectively.
- Volkswagen is priced more in the mid-range, at 5402
- The rest of the top 9 brands are all priced more moderately around 2000-4000.

## Exploring mileage of top brands
Previously, we've isolated the most popular brands and looked at mean prices for these. We'll now look at mileage for those same brands, and explore a potential link to mean prices.

In [27]:
mean_mileages = {}

for curr_brand in popular_brands:
    brand_mileages = autos.loc[autos['brand'] == curr_brand, 'odometer_km']
    mean_mileages[curr_brand] = brand_mileages.mean()

mean_mileages

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056,
 'renault': 128071.33121308497,
 'peugeot': 127153.62526920316,
 'fiat': 117121.9715956558}

Let's convert both the mean_mileage and mean_price dictionaries to series to then combine them in a single dataframe:

In [28]:
mean_mileages_ser = pd.Series(mean_mileages)
mean_price_ser = pd.Series(mean_prices)

brand_info = pd.DataFrame(mean_mileages_ser, columns=['mean_mileage_km'])
brand_info

mean_mileage_km
volkswagen       128707.158791
bmw              132572.513140
opel             129310.035842
mercedes_benz    130788.363313
audi             129157.386785
ford             124266.012872
renault          128071.331213
peugeot          127153.625269
fiat             117121.971596

In [30]:
brand_info['mean_price_dollar'] = mean_price_ser
brand_info = brand_info.sort_values('mean_mileage_km', ascending=False)
brand_info

mean_mileage_km  mean_price_dollar
bmw              132572.513140               8332
mercedes_benz    130788.363313               8628
opel             129310.035842               2975
audi             129157.386785               9336
volkswagen       128707.158791               5402
renault          128071.331213               2474
peugeot          127153.625269               3094
ford             124266.012872               3749
fiat             117121.971596               2813

The mileages are fairly evenly distributed, as all of the popular brands have a mileage of between 115,000 and 135,000 km. The table seems to suggest that higher mileage brands corresponds to higher priced brands. This could be a result of the more expensive cars being more durable, thus making higher mileages possible. If one has a more exclusive car one might also be motivated to keep it in good shape, increasing its life expectancy.